rules/no-complex-iteratee-expression.js

```diff
@@ -0,0 +1,115 @@
+'use strict';
+const {flatten} = require('lodash');
+const pluralize = require('pluralize');
+const avoidCapture = require('./utils/avoid-capture');
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+const MESSAGE_ID = 'no-complex-iteratee-expression';
+const messages = {
+	[MESSAGE_ID]: 'Move the complex iteratee expression out of the "for-of" header.'
+};
+
+const complexForSelector = [
+	'ForOfStatement',
+	`:not(${
+		[
+			'[right.type="Identifier"]',
+			'[right.type="MemberExpression"]',
+			'[right.type="CallExpression"][right.arguments.length=0]',
+			[
+				'[right.type="CallExpression"]',
+				'[right.callee.type="MemberExpression"]',
+				'[right.callee.object.type="Identifier"]',
+				'[right.callee.object.name="Object"]',
+				'[right.callee.property.type="Identifier"]',
+				'[right.callee.property.name=/(keys|values|entries)/]'
+			].join('')
+		].join(', ')
+	})`
+].join('');
+
+const getScopes = scope => [
+	scope,
+	...flatten(scope.childScopes.map(scope => getScopes(scope)))
+];
+
+const getIndentString = (node, sourceCode) => {
+	const {line, column} = sourceCode.getLocFromIndex(node.range[0]);
+	const lines = sourceCode.getLines();
+	const before = lines[line - 1].slice(0, column);
+
+	return before.match(/\s*$/)[0];
+};
+
+const scopeStatements = new Set(['DoWhileStatement', 'ForInStatement', 'ForOfStatement', 'ForStatement', 'WhileStatement', 'WithStatement']);
+const shouldAddBraces = node => {
+	if (node.parent.type === 'IfStatement') {
+		const nodeType = node.parent.consequent === node ? 'consequent' : 'alternate';
+		return node.parent[nodeType].type !== 'BlockStatement';
```

Comment:
Do not check type, check `parent.consequent === node`, not the same thing.

Reply:
I don't see why, because if the for loop is in the alternate, we want to format the alternate. So if you check whether it is the consequent, we will miss the case where it is in an `else`.
Am I right?

Reply:
Your code works fine in current esree spec, only because the `IfStatement` can't be other child of the listing parents.

But imagine if your checking node is an `Identifier`, your test will all wrong, because if can be other child, safest way is to check `parent.xxx === node` to make sure it is correct child of parent. Even if the spec add other child.

Reply:
Oh ok. So I should just return `node.parent.consequent === node`? Should I do the same for `SwitchCase`, or only for the `If`?

Reply:
All of them.  Just use

```js
return node.parent.type === "XXX" && node.parent.yyy ==== node
```

Means node is parent with type `XXX`'s property `yyy`